elements/lisk-bft/src/finality_manager.ts

```diff
@@ -411,27 +411,17 @@ export class FinalityManager extends EventEmitter {
 			CONSENSUS_STATE_VALIDATOR_LEDGER_KEY,
 		);
 		const { ledger } = this._decodeVotingLedger(bftState);
-		const { preVoted } = this._getChainMaxHeightStatus(ledger);
-
-		return preVoted;
+		return this._calculateMaxHeightPrevoted(ledger);
 	}
 
-	private _getChainMaxHeightStatus(ledger: LedgerMap): { preVoted: number; preCommitted: number } {
+	private _calculateMaxHeightPrevoted(ledger: LedgerMap): number {
 		debug('updatePreVotedAndFinalizedHeight invoked');
 
-		const highestHeightPreVoted = Object.keys(ledger)
+		const maxHeightPreVoted = Object.keys(ledger)
 			.reverse()
 			.find(key => ledger[key].preVotes >= this.preVoteThreshold);
 
-		const preVoted = highestHeightPreVoted ? parseInt(highestHeightPreVoted, 10) : 0;
-
-		const highestHeightPreCommitted = Object.keys(ledger)
-			.reverse()
-			.find(key => ledger[key].preCommits >= this.preCommitThreshold);
-
-		const preCommitted = highestHeightPreCommitted ? parseInt(highestHeightPreCommitted, 10) : 0;
-
-		return { preVoted, preCommitted };
+		return maxHeightPreVoted ? parseInt(maxHeightPreVoted, 10) : this.finalizedHeight;
```

Comment:
I don't fully understand the code:
- Why is `maxHeightPreVoted` a string and needs to be pared and not a number?
- Do I return this.finalizedHeight if maxHeightPreVoted is undefined?
- The `finalizedHeight` is part of the state and initialized with the height of the genesis block, right? The `maxHeightPrevoted` is not part of the state and always computed from scratch?
- In the prototype implementation `maxHeightPrevoted` together with `finalizedHeight` is maintained as part of the state. The update works as follows: Go over the last 303 block headers and checking and if for any of them the prevotes/precommits and update if the height is larger. If you only store the `finalizedHeight` explicitly, you will have to go over all block headers from `finalizedHeight` onwards (this number can be arbitrarily large, e.g., 10,000). Does this happen here? 
-

Reply:
Cosmetic remark: The camelCase is a bit inconsistent, sometimes prevote and sometimes preVote. Note that the LIP uses prevote and precommit as one work in lower case.

Reply:
Another remark: Why are there two methods `getMaxHeightPrevoted` and `_calculateMaxHeightPrevoted`? The second method is called directly as well, but before the call it seems the same things happens as in  `getMaxHeightPrevoted`.

Reply:
> Why is maxHeightPreVoted a string and needs to be parsed and not a number?

Node.js object key is always a string even though we assign a number. Therefore, it needs to be parsed.

> Do I return this.finalizedHeight only if maxHeightPreVoted is undefined? Does this happen only if there is no block which reaches the prevoteThreshold?

yes. maxHeightPrevoted is undefined only if there is no block height which is above prevoteThreshold.

> The finalizedHeight is part of the state and initialized with the height of the genesis block, right? The maxHeightPrevoted is not part of the state and always computed from scratch?

> In the prototype implementation maxHeightPrevoted together with finalizedHeight is maintained as part of the state. The update works as follows: Go over the last 303 block headers, check if for any of them the prevotes/precommits threshold is reach and update if the height is larger. If you only store the finalizedHeight explicitly, you will have to go over all block headers from finalizedHeight onwards (this number can be arbitrarily large, e.g., 10,000). Does this happen here?

prevotes/precommits are also stored as state until finalized height. it's called `ledger` in the code. the structure is like
```
{
  ${height}: {
    preVotes: number,
    preCommits: number,
  }
}
```
so moving back and forward is managed along with account state.

Reply:
`_calculateMaxHeightPrevoted` is used in 2 different places, so that's why it is extracted to private method